File: solve/cpp/radix.cc
#include "common.hh"

void arrange(int *beg, int *end, int *tar, int base, int digit)
{
    for (int r = 0; r != base; ++r)
        for (int *p = beg; p != end; ++p)
            if (((*p) / digit) % base == r)
                *tar++ = *p;
} 

void radixsort(int base)
{
    int digit = 1;
    int *b = new int[n];
    for (int i = 0; i != 8; ++i)
    {
        if (i%2 == 0)
            arrange(a, a + n, b, base, digit);
        else
            arrange(b, b + n, a, base, digit);
        digit *= base;
    }
    delete b;
}

int main(int argc, char *argv[])
{
    if (init(argc, argv) == 0)
    {
        radixsort(10);
        output();
        return 0;
    }
    return 1;
}

File: solve/lua/radix.lua
require "common" 

function arrange(from, to, base, digit)
    local k = 1
    for r = 0, base - 1 do
        for i = 1, n do
            if math.floor(from[i]/digit)%base == r then
                to[k] = from[i]
                k = k + 1
            end 
        end     
    end     
end

function radixsort(base)
    local b = {  }
    local digit = 1
    for i = 1, 8 do
        if i%2 == 0 then
            arrange(b, a, base, digit)
        else
            arrange(a, b, base, digit)
        end
        digit = digit*base
    end     
end

skeleton(function ()
    radixsort(10)
end)
